Adding Item and Sword class

CamYuan opened the pull request:
# **Overview**
Adding General Item class and Sword class which the player can pick up and use as a weapon to kill DontTouchMe enemies. Sword hitbox animation is managed by the HitboxAnimationManager. Hitbox locations, angles, and frame durations are passed to the hitbox manager and can be activated to 'animate' through each hitbox.

# **Features**
## *Items* - New General Class
* Some items are wieldable and can be equipped by the player
* Equipped items can be dropped by the player
* Weapon items have an attack_state
* Any equipped item keeps track of the player as a reference point
## *Sword* - New Class Extends Item
* Weapon specific collision function
* Reset the item after attack_state is complete
* Create custom hitbox frames for attack_state animation
## *Player* - Added methods
* Added functionality to pickup and drop items
* Key 'SPACE' is used to attack if the player is holding a weapon item
* Key 'g' is used to drop an item currently being wielded
* Player keeps track if he is wielding an item and what that item is
* Player keeps track of whether or not he is in the attack state
## *Playstate* - Added Sprites
* Added sword sprite
* Added sword collision logic to the playstate
## *HitboxAnimationManager* - New Control Class
* Creates hitboxes based on input offsets, angles and stores how many frames each hitbox should show
* Animates through the hitboxes and shows in the appropriate frames


# **Testing**
* Items should collide with the map and blocks as normal.
* Items should be picked up by the player by colliding with them
* Items can be dropped by the player by hitting the 'g' key
* Items just dropped should not be able to be picked up immediately
* Dropped items should act as normal and collide with the map and blocks
* Dropped items should not fall through the map
* If an item is equipped, player can attack with 'space'
* Player should not be able to attack if not wielding an item
* Items cannot be dropped while in the attack state
* While attacking, If a hitbox collides with a DontTouchMe enemy, the enemy should die
* Enemies should not die if the player is not in the attack state

Commit:
-added functionality to keep track of the direction the player is facing.
- Added keys to attack if a weapon is equipped
- Added key to drop a weapon if it is equipped
- Added player ability to pick up an item
- Overloaded the kill to reset the game after 3 seconds of the player dying. This can be removed later if unwanted

Commit:
- Added sword to playstate
- Added collision for the sword against the map, player, and enemies

Commit:
Item class
-General item class allowing the player to interact with items.
-item keeps track of the player if equipped
-item can be dropped
-item can be used to attack if it is weildable and equipped

Commit:
Sword class and image
- Overload attack function for weapon specific attack
- attack functionality is still in the works

Commit:
Item.hx
-Added comments and formatted lines
-Removed some unnecessary code
-Added justDropped bool and timer to properly drop  and delay pickup
-Adjusted drop logic to be smoother

Commit:
Player.hx
- Removed the Item Bag. This could be a future functionality, but it is not used right now
- added comments and formatted lines
- Created a function for dropping an item and called it in the appropriate places
- Adjusted pickup logic to allow delay if an item was just dropped

Commit:
Playstate.hx
- Added the sword hittbox and hitbox frames to the playstate.
- Changed the overlap so that it only checks the frames ( which only appear during attack animation)
- Enable keys on the reset of the level. (If the player was dying while in the attack state, keys stay disabled)

Commit:
Sword.hx
- Removed unnecessary code
- Added comments and formatted lines
- Added code to create a frameAnimationManager
- Now tracks the player and faces the same way as the player

Commit:
Class FrameAnimationManager
- Creates hitboxes for animating objects and loops through for a set amount of frames from the input.

Commit:
Renamed FrameAnimationManager to HitboxAnimationManager because thats more inline with what it is

Commit:
Delete FrameAnimationManager.hx

Delete the old name of the file

Commit:
Merge branch 'master' into Items

Cortillaen commented on source/Player.hx:
```diff
@@ -187,6 +210,48 @@ import flixel.group.FlxGroup;
+		weilding = false;
+		equipped_item = null;
+	}
+}
```
If `bounce()` is intended to be a method of the `Player` class, I think you need to remove this close-bracket.  Currently, it won't build since `public function bounce()` doesn't make sense unless it is part of a class (and the extra close-bracket afterward, but the build stops because of `public` without ever getting there).

CamYuan replied:
That was a mistake on my part when resolving merge conflicts. I accidentally left the extra curly brace that ends the player class

Cortillaen commented on source/Sword.hx:
```diff
@@ -0,0 +1,84 @@
+	 * @param	sword
+	 * @param	enemy
+	 */
+	public function hit_enemy(sword:Sword, enemy:DontTouchMe){
```
Is there any reason to hard-code the sword to only work on one enemy type?  If not, it might be better to use the `Enemy` class and have it call a function that each sub-class defines to handle any unique behavior when hit with the sword.

CamYuan replied:
That is correct. The behavior of the hit_enemy could be unique depending on the item, but it should collide with several enemies. In the current implementation, there is only one type of enemy and only one of it. Optimally I would collide the sword with a Group of enemies

Cortillaen commented on source/Sword.hx:
```diff
@@ -0,0 +1,84 @@
+import flixel.system.FlxAssets.FlxGraphicAsset;
+
+/**
+ * SWORDS!!!!!!
```
[Okay, Fighter.](http://static.tvtropes.org/pmwiki/pub/images/cit_8_bit_theatre_-_welcome_to_corneria_-_urge_to_kill_rising.png)

Cortillaen commented on source/PlayState.hx:
```diff
@@ -168,13 +173,19 @@ class PlayState extends FlxState
 		FlxG.collide(map, dtmEnemy1);
 		FlxG.collide(map, bullets);
 		FlxG.collide(blockGroup, bullets);
+		if(player.equipped_item != sword){
```
This works for the moment, but it will have to be overhauled if more than one sword is in a level or other items are introduced.  It might be better to add all items to a group, run collide on that, and remove items from the group when they are picked up (making sure to add them back when dropped).

CamYuan replied:
This is temporary since I only had one item implemented and only one sword in the level. In the future, this would be changed to accommodate several items or Groups of items

Cortillaen commented on source/Item.hx:
```diff
@@ -0,0 +1,86 @@
+	 * weilded. Disables user input to put them in the 'attacking state', but 
+	 * still keeps their current velocity and acceleration
+	 */
+	public function attack_state(){
```
Currently, this makes any attack cause you to stop moving (you retain velocity and accel, but not holding a movement key, which this mimics, causes you to stop almost instantly).  Is there a better way than simply locking user input during the attack animations?

Also, is this really something that belongs in `Item`?  It purely affects input to `Player`, not this class.

CamYuan replied:
This was the most time efficient solution for me. There could be a better way if I created a custom attack_state and added it to the player states folder. Unfortunately, I did not want to mess with too much of the FSM and a simple bool check and disabling the keys was an easy way to achieve my functionality.

CamYuan replied:
Also the state was moved into the player Class

Cortillaen commented on source/Item.hx:
```diff
@@ -0,0 +1,86 @@
+ */
+class Item extends FlxSprite
+{
+	private var player_trace:Player;
```
Keeping an internal copy of `Player` feels really awkward.  If it is just to access things like player facing, could you maybe do that through `FlxG` instead?

CamYuan replied:
There are several times where the position of the player, as well as the way the player is facing needs to be known by the weapon in order to have any animations in the correct direction, angles, x/y positions in relation to the player. I modeled the tracking off of the Sentry Enemy. Do you have another suggestions to keep track of these?

Cortillaen replied:
For the direction, that could just be passed into some function of either `Item` or `Sword` whenever the player moves left or right.  Something like 
`public function setFacing(newFacing:Int) {
  facing = newFacing;
}`
and use that in `Player` by replacing the current `facing = FlxObject.LEFT;` (and the right one) with
`if (weilding) {
  equipped_item.setFacing(FlxObject.LEFT);
}`

You should be able to access the `Player`'s coordinates from inside `Item` or `Sword` (I'm leaning toward the former) with `(cast (FlxG.state, PlayState)).player.x`  You could also leave the facing part of `Player` as is and just access it this way.

CamYuan replied:
On a side note, implementation could be changed to keep track of an FlxObject and keep the tracker within the class. This would allow multiplayer item memory as well as allowing enemies to have specific Items. I will do as you suggested, but this is a potential implementation

Cortillaen commented on source/HitboxAnimationManager.hx:
```diff
@@ -0,0 +1,125 @@
+package;
```
This whole class may be more trouble than it's worth since it doesn't appear to be working as intended.  Currently, the hitbox for the sword does not rotate along with the graphic.
![image](https://cloud.githubusercontent.com/assets/14339030/26087510/c315c448-39b6-11e7-81db-a3439a35c341.png)
It might be better (simpler, at least, and reasonably close to the intended functionality) to simply resize the hitbox during the swing animation to a rectangle that covers the whole area of the swing.

CamYuan replied:
Nice Catch and thanks for the screenshot. I didn't realize this was still happening with them being FlxSprites. I noticed it when I had them as FlxObjects and talked to Sam about it. He suggested having multiple hitboxes of different sizes stacked on each other in order to get the angled effect. I'll work on getting this fixed tomorrow

Cortillaen replied:
Fun thing I learned about Github with that:  You can paste a screenshot straight from the clipboard into a comment like this, and it'll upload and insert it automatically.  QoL right there.

Also, I'm not sure why I didn't note this before, but the sword image on the player should probably go `visible = false` during the swing animation.

Cortillaen commented on source/PlayState.hx:
```diff
@@ -168,13 +173,19 @@ class PlayState extends FlxState
 		FlxG.collide(blockGroup, bullets);
+		if(player.equipped_item != sword){
+			FlxG.collide(map, sword);
+			FlxG.collide(blockGroup, sword);
```
These will have to be rewritten in the event that any other items are introduced or there are multiple swords in a level.  It might be better to put all items in a group and run collisions on that.  If there are issues with collisions when an item is being carried, it could just be pulled out of the group when picked up and added back when dropped.

CamYuan replied:
Correct. I only added one sword to the level, but if there were more, they would be put into a Group and checked the collision there.

Cortillaen commented on source/Sword.hx:
```diff
@@ -0,0 +1,84 @@
+	 */
+	private function reset_weapon(){
+	  player_trace.attacking = false;
+	  FlxG.keys.enabled = true;
```
See above comment on locking input.

CamYuan replied:
I've change this to s simple bool check. Although it now allows to player to switch directions and attack very rapidly. Working on a solution for these things now

Cortillaen replied:
At the moment, hitting space starts an infinite loop of attacks and prevents any further input, so I assume you're still working on it.

Cortillaen commented on source/Player.hx:
```diff
@@ -187,6 +210,48 @@ import flixel.group.FlxGroup;
+		weilding = false;
+		equipped_item = null;
+	}
+}
```
This bracket needs removed.  It currently closes the class before getting to `bounce()`

Cortillaen commented on source/PlayState.hx:
```diff
@@ -204,5 +215,6 @@ class PlayState extends FlxState
 	private function resetLevel(Timer:FlxTimer):Void
 	{
 		FlxG.resetState();
+		FlxG.keys.enabled = true;
```
See comment on locking input above.

Cortillaen commented on source/Item.hx:
```diff
@@ -0,0 +1,86 @@
+	 */
+	public function attack_state(){
+		FlxG.keys.reset();
+		FlxG.keys.enabled = false;
```
What is the intent on this?  Disabling all user input when attacking seems excessive and cumbersome.  Even though velocity is maintained, acceleration is not since it defaults to just gravity unless a key is pressed.  If you just need to prevent starting a new swing until the current one is done, it might be better to just use a bool to indicate being in an attack and just check that when processing key-presses.

Also, should this even be in `Item`?  It affects player input, which is handled in `Player`.

CamYuan replied:
The idea was to stop the player from moving around for a second (but still keeping momentum) while he attacks. I believe this is purely an opinion based idea on how a game should work. If I allow keys to be enabled. Users could switch left and right rapidly to be 'attacking' on both sides since the animation will flip.

Cortillaen commented on source/Player.hx:
```diff
@@ -187,6 +210,48 @@ import flixel.group.FlxGroup;
+	  new FlxTimer().start(2, resetLevel, 1); //This helps speed things up for debugging
+  }
+  
+  private function resetLevel(Timer:FlxTimer):Void
```
I'm not sure an operation like this really belongs in `Player`, though moving it outside is basically just adding an extra bool and check.

CamYuan replied:
It felt more out of place in the Playstate, with the extra logical check since this is something that doesn't need one if it's called here in the Player. I can move the logic into the Playstate if necessary, but I still think it fits best here.

Cortillaen replied:
Might poke @SamBumgardner about it.  I know it would require the extra check to put it in `PlayState`, but it's literally a State operation.

CamYuan replied:
Moved the function itself to the playstate, but I still kept the override in the player class. I access the playstate method through FlxG the way you suggested I access the player class earlier.

Cortillaen commented on source/Item.hx:
```diff
@@ -0,0 +1,86 @@
+ */
+class Item extends FlxSprite
+{
+	private var player_trace:Player;
```
Is keeping a copy of `Player` here necessary?  What about having `Player` update the equipped item through methods or, if direct access is absolutely necessary, just accessing the `Player` object through FlxG?

CamYuan replied:
I didn't know I could do that. Could you give me an example of how I can reference the Player object through FlxG?

Cortillaen replied:
I left an example above.

Cortillaen reviewed (changes requested):
Overall, it's looking good.  I really appreciate the documentation, too.

There is currently one typo that prevents building (the extra close-bracket before `bounce()` in `Player`) and a few things I wonder if should be reworked.  All of the functionality appears to be there, though, so nice work!

Commit:
Code Fixes

Cortillaen commented on source/Player.hx:
```diff
@@ -187,6 +210,58 @@ import flixel.group.FlxGroup;
+	 */
+	public function attack_state():Bool{
+		FlxG.keys.reset();
+		FlxG.keys.enabled = false;
```
Not sure what happened, but one of my comments disappeared.  All the comments about locking input were supposed to be referring to this one.

This just feels bad to me.  It causes you to stop moving during your attack animation (you keep velocity, but not holding any movement key causes the player to come to a quick stop, and the `reset()` cancels any keys being held) and just makes it feel more clunky than I think is necessary.  If the intention here is just to prevent starting a new swing before the previous one is complete, I think just using a Bool flag and logic check in `Player` is a better choice than locking all input during the animation.

Commit:
Moving some code around

Cortillaen commented on source/PlayState.hx:
```diff
@@ -204,5 +215,6 @@ class PlayState extends FlxState
 	private function resetLevel(Timer:FlxTimer):Void
```
I just noticed this will need to be a public to access through FlxG.

Cortillaen reviewed (changes requested):
There seems to be a new issue where hitting the attack key starts an infinite loop of attacks and prevents any further input.

Commit:
Changed some logic to make gameplay more appealing

Commit:
Finished taking aplyer_trace out of the item and it now references the playerstate.player object.
Finished fixing the sword attack hitboxes, so they extend and retract, but it's still not perfect.
Removing a file I accidentally added

Commit:
Accidentally left a debug statement in

Commit:
Added sprites to the animation so that people can visualize it more easily

Commit:
Merge branch 'master' into Items

Commit:
Changed the attack to to be on release. This helps the multiple attack gltich appear less often

Commit:
Merge branch 'Items' of https://github.com/CamYuan/csc-303-platformer into Items
